File: repository-local/src/main/java/com/andyanika/translator/repository/local/LocalRepositoryImpl.java
package com.andyanika.translator.repository.local;

import android.content.SharedPreferences;

import com.andyanika.translator.common.interfaces.LocalRepository;
import com.andyanika.translator.common.models.FavoriteModel;
import com.andyanika.translator.common.models.LanguageCode;
import com.andyanika.translator.common.models.TranslateDirection;
import com.andyanika.translator.common.models.TranslateRequest;
import com.andyanika.translator.common.models.TranslateResult;
import com.andyanika.translator.repository.local.model.WordModel;

import java.util.List;

import io.reactivex.Flowable;
import io.reactivex.Observable;
import io.reactivex.Scheduler;
import io.reactivex.Single;
import io.reactivex.subjects.PublishSubject;
import timber.log.Timber;

class LocalRepositoryImpl implements LocalRepository {
    static final String LANGUAGE_SRC = "language_src";
    static final String LANGUAGE_DST = "language_dst";


    private final TranslatorDao dao;
    private final SharedPreferences preferences;
    private final ModelsAdapter adapter;
    private final Scheduler ioScheduler;

    private final PublishSubject<LanguageCode> srcLanguageSubject = PublishSubject.create();
    private final PublishSubject<LanguageCode> dstLanguageSubject = PublishSubject.create();

    LocalRepositoryImpl(TranslatorDao dao, SharedPreferences preferences, ModelsAdapter adapter, Scheduler ioScheduler) {
        this.dao = dao;
        this.preferences = preferences;
        this.adapter = adapter;
        this.ioScheduler = ioScheduler;
    }

    @Override
    public Flowable<List<FavoriteModel>> getHistory() {
        return dao.getHistory()
                .flatMap(list -> Flowable.fromIterable(list)
                        .map(adapter::toTranslationRowModel).toList()
                        .toFlowable());
    }

    @Override
    public Flowable<List<FavoriteModel>> getFavorites() {
        return dao.getFavorites()
                .flatMap(list -> Flowable.fromIterable(list)
                        .map(adapter::toTranslationRowModel).toList()
                        .toFlowable());
    }

    @Override
    public Single<TranslateResult> translate(TranslateRequest request) {
        return dao.getTranslation(request.getText(), request.getDirection().getSrc().toString(), request.getDirection().getDst().toString())
                .map(adapter::toTranslationResult);
    }

    @Override
    public void addTranslation(TranslateResult translateResult) {
        WordModel model = adapter.toWordModel(translateResult);
        dao.addTranslation(model);
    }

    @Override
    public void addFavorites(int wordId) {
        com.andyanika.translator.repository.local.model.FavoriteModel model = new com.andyanika.translator.repository.local.model.FavoriteModel(wordId);
        dao.addFavorite(model);
    }

    @Override
    public void removeFavorite(int wordId) {
        com.andyanika.translator.repository.local.model.FavoriteModel model = new com.andyanika.translator.repository.local.model.FavoriteModel(wordId);
        dao.removeFavorite(model);
    }

    @Override
    public Observable<LanguageCode> getSrcLanguage() {
        return srcLanguageSubject
                .startWith(Observable.fromCallable(() -> {
                            String s = preferences.getString(LANGUAGE_SRC, null);
                            return LanguageCode.tryParse(s, LanguageCode.RU);
                        }
                ))
                .onErrorReturnItem(LanguageCode.RU)
                .subscribeOn(ioScheduler);
    }

    @Override
    public Observable<LanguageCode> getDstLanguage() {
        return dstLanguageSubject
                .startWith(Observable.fromCallable(() -> {
                            String s = preferences.getString(LANGUAGE_DST, null);
                            return LanguageCode.tryParse(s, LanguageCode.EN);
                        }
                ))
                .onErrorReturnItem(LanguageCode.EN)
                .subscribeOn(ioScheduler);
    }


    @Override
    public void setLanguageDirection(TranslateDirection<LanguageCode> direction) {
        Timber.d("save repository direction: %s - %s", direction.getSrc(), direction.getDst());
        preferences
                .edit()
                .putString("language_src", direction.getSrc().toString())
                .putString("language_dst", direction.getDst().toString())
                .apply();

        srcLanguageSubject.onNext(direction.getSrc());
        dstLanguageSubject.onNext(direction.getDst());
    }

    @Override
    public Observable<LanguageCode> getAvailableLanguages() {
        return Observable.fromArray(LanguageCode.values()).cache();
    }
}

File: repository-local/src/main/java/com/andyanika/translator/repository/local/LocalRepositoryImpl.kt
package com.andyanika.translator.repository.local

import android.content.SharedPreferences
import com.andyanika.translator.repository.local.entity.FavoriteEntity
import core.interfaces.LocalRepository
import core.models.*
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.withContext
import timber.log.Timber

internal class LocalRepositoryImpl(
    private val dao: TranslatorDao,
    private val preferences: SharedPreferences,
    private val adapter: ModelsAdapter,
    private val ioDispatcher: CoroutineDispatcher,
) : LocalRepository {

    private val srcLanguageState = MutableStateFlow(getSrcLanguage())
    private val dstLanguageState = MutableStateFlow(getDstLanguage())

    override fun getHistory(): Flow<List<FavoriteModel>> {
        return dao
            .getHistory()
            .map { models -> models.map(adapter::toTranslationRowModel) }
    }

    override fun getFavorites(): Flow<List<FavoriteModel>> {
        return dao
            .getFavorites()
            .map { models -> models.map(adapter::toTranslationRowModel) }
    }

    override suspend fun translate(request: TranslateRequest): TranslateResult? {
        return withContext(ioDispatcher) {
            dao.getTranslation(
                text = request.text,
                langSrc = request.direction.src.toString(),
                langDst = request.direction.dst.toString()
            )?.let(adapter::toTranslationResult)
        }
    }

    override suspend fun addTranslation(translateResult: TranslateResult) {
        return withContext(ioDispatcher) {
            val model = adapter.toWordModel(translateResult)
            dao.addTranslation(model)
        }
    }

    override suspend fun addFavorites(wordId: Int) {
        withContext(ioDispatcher) {
            val entity = FavoriteEntity(wordId)
            dao.addFavorite(entity)
        }
    }

    override suspend fun removeFavorite(wordId: Int) {
        withContext(ioDispatcher) {
            val entity = FavoriteEntity(wordId)
            dao.removeFavorite(entity)
        }
    }

    override fun observeSrcLanguage(): Flow<LanguageCode> {
        return srcLanguageState
    }

    override fun observeDstLanguage(): Flow<LanguageCode> {
        return dstLanguageState
    }

    override fun getSrcLanguage(): LanguageCode {
        val s = preferences.getString(LANGUAGE_SRC, null)
        return LanguageCode.tryParse(s, LanguageCode.RU)
    }

    override fun getDstLanguage(): LanguageCode {
        val s = preferences.getString(LANGUAGE_DST, null)
        return LanguageCode.tryParse(s, LanguageCode.EN)
    }

    override suspend fun setLanguageDirection(direction: TranslateDirection<LanguageCode>) {
        Timber.d("save repository direction: %s - %s", direction.src, direction.dst)
        preferences
            .edit()
            .putString("language_src", direction.src.toString())
            .putString("language_dst", direction.dst.toString())
            .apply()
        srcLanguageState.emit(direction.src)
        dstLanguageState.emit(direction.dst)
    }

    override suspend fun getAvailableLanguages(): List<LanguageCode> {
        return LanguageCode.values().toList()
    }

    companion object {
        const val LANGUAGE_SRC = "language_src"
        const val LANGUAGE_DST = "language_dst"
    }
}
